plugin/src/main/scala/chisel3/internal/plugin/BundleComponent.scala

```diff
@@ -189,6 +194,43 @@ private[plugin] class BundleComponent(val global: Global, arguments: ChiselPlugi
       elementsImpl
     }
 
+    def generateAutoTypename(bundle: ClassDef, thiz: global.This): Option[Tree] = {
+      val (con, params) = getConstructorAndParams(bundle.impl.body, true)
+      if (con.isEmpty) {
+        global.reporter.warning(bundle.pos, "Unable to determine primary constructor!")
+        return None
+      }
+
+      val constructor = con.get
+      // The params have spaces after them (Scalac implementation detail)
+      val paramLookup: String => Symbol = params.map(sym => sym.name.toString.trim -> sym).toMap
+
+      // Create a this.<ref> for each field matching order of constructor arguments
+      // Unlike autoCloneType, just a list of Tree since we only care about the parameters themselves and not parameter lists
+      val typeNameConArgs: List[(Tree)] =
+        constructor.vparamss.flatMap(_.map { vp =>
+          val p = paramLookup(vp.name.toString)
+          // Make this.<ref>
+          val select = gen.mkAttributedSelect(thiz.asInstanceOf[Tree], p)
+          // Clone any Data parameters to avoid field aliasing, need full clone to include direction
+          val cloned = if (isData(vp.symbol)) cloneTypeFull(select.asInstanceOf[Tree]) else select
+          // Need to splat varargs
+          if (isVarArgs(vp.symbol)) q"$cloned: _*" else cloned
+        })
```

Comment:
I think all of this is basically copy-pasted from `generateAutoCloneType`, please figure out how to refactor the code such that the code is not duplicated and so that it only runs once.